hbase-balancer/src/main/java/org/apache/hadoop/hbase/CatalogAccessor.java

```diff
@@ -0,0 +1,1071 @@
+      // Ignore. This is used with tableName passed as regionName.
+    }
+    Get get = new Get(row);
+    get.addFamily(HConstants.CATALOG_FAMILY);
+    Result r;
+    try (Table t = getCatalogHTable(connection, getCatalogTableForTable(parsedInfo.getTable()))) {
+      r = t.get(get);
+    }
+    RegionLocations locations = CatalogFamilyFormat.getRegionLocations(r);
+    return locations == null ? null :
+      locations.getRegionLocation(parsedInfo == null ? 0 : parsedInfo.getReplicaId());
+  }
+
+  /**
+   * Returns the HRegionLocation from meta for the given region
+   * @param connection connection we're using
+   * @param regionInfo region information
+   * @return HRegionLocation for the given region
+   */
+  public static HRegionLocation getRegionLocation(Connection connection, RegionInfo regionInfo)
+    throws IOException {
+    return CatalogFamilyFormat.getRegionLocation(getCatalogFamilyRow(connection, regionInfo),
+      regionInfo, regionInfo.getReplicaId());
+  }
+
+  /**
+   * @return Return the {@link HConstants#CATALOG_FAMILY} row from hbase:meta.
+   */
+  public static Result getCatalogFamilyRow(Connection connection, RegionInfo ri)
+    throws IOException {
+    Get get = new Get(CatalogFamilyFormat.getMetaKeyForRegion(ri));
+    get.addFamily(HConstants.CATALOG_FAMILY);
+    try (Table t = getCatalogHTable(connection, getCatalogTableForTable(ri.getTable()))) {
+      return t.get(get);
+    }
+  }
+
+  public static TableName getCatalogTableForTable(TableName tableName) {
+    if (TableName.ROOT_TABLE_NAME.equals(tableName)) {
+      throw new IllegalStateException("Can't get catalog table for hbase:root table");
+    }
+    if (TableName.META_TABLE_NAME.equals(tableName)) {
+      return TableName.ROOT_TABLE_NAME;
+    }
+    return TableName.META_TABLE_NAME;
+  }
+
+  public static boolean isCatalogTable(TableName tableName) {
+    return tableName.equals(TableName.ROOT_TABLE_NAME) ||
+      tableName.equals(TableName.META_TABLE_NAME);
+  }
+
+  /**
+   * Gets the result in hbase:meta for the specified region.
+   * @param connection connection we're using
+   * @param regionName region we're looking for
+   * @return result of the specified region
+   */
+  public static Result getRegionResult(Connection connection,
+      byte[] regionName) throws IOException {
+    Table catalogTable = null;
+    if (Bytes.equals(RegionInfoBuilder.ROOT_REGIONINFO.getRegionName(), regionName)) {
+      throw new IllegalStateException("This method cannot be used for hbase:root region");
+    }
+    catalogTable = getCatalogHTable(connection,
+      getCatalogTableForTable(
+        CatalogFamilyFormat.parseRegionInfoFromRegionName(regionName).getTable()));
+    Get get = new Get(regionName);
+    get.addFamily(HConstants.CATALOG_FAMILY);
+    try (Table t = catalogTable) {
+      return t.get(get);
+    }
+  }
+
+  /**
+   * Scans META table for a row whose key contains the specified <B>regionEncodedName</B>, returning
```

Comment:
Only hbase:meta?

Reply:
Looks like it does both if nothing found in hbase:meta. Worth mention in javadoc.

Reply:
Will do.